pkg/cmd/cli/cmd/startbuild.go

```diff
@@ -433,17 +447,35 @@ func streamPathToBuild(repo git.Repository, in io.Reader, out io.Writer, client
 		}
 		fmt.Fprintf(out, "Uploading archive file from STDIN as binary input for the build ...\n")
 
-	default:
-		var fromPath string
-		switch {
-		case asDir:
-			fromPath = fromDir
-		case asFile:
-			fromPath = fromFile
-		case asRepo:
-			fromPath = fromRepo
+	case (asFile || asDir) && (strings.HasPrefix(fromPath, "http://") || strings.HasPrefix(fromPath, "https://")):
+		resp, err := http.Get(fromPath)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != 200 {
+			return nil, fmt.Errorf("unable to download file %q: %s", fromPath, resp.Status)
 		}
 
+		r = resp.Body
+
+		if asFile {
+			options.AsFile = httpFileName(resp)
+			if options.AsFile == "" {
+				return nil, fmt.Errorf("unable to determine filename from HTTP headers or URL")
+			}
+			fmt.Fprintf(out, "Uploading file from %q as binary input for the build ...\n", fromPath)
+		} else {
+			br := bufio.NewReaderSize(resp.Body, 4096)
+			r = br
+			if !isArchive(br) {
+				fmt.Fprintf(out, "WARNING: the provided file may not be an archive (tar, tar.gz, or zip), use --from-file\n")
```

Comment:
maybe we just need to extend the warning (in all locations) to something like "use --as-file to prevent extraction"